### PR TITLE
WIP: test/e2e/storage: replace mock driver with hostpath driver

### PR DESCRIPTION
**Pull request by avalluri**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This is a first step towards removing the mock CSI driver completely from
e2e testing in favor of hostpath plugin. With the recent hostpath plugin
changes(PR #260, #269), it supports all the features supported by the mock
csi driver.

Using hostpath-plugin for testing also covers CSI persistent feature
usecases.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
The PR is in WIP as it requires a hostpath-plugin release with the needed changes. Currently, it is using hostpath 'canary' images.

#### Does this PR introduce a user-facing change?
No

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml [hunk 1]
```diff
@@ -93,3 +103,12 @@ spec:
             path: /var/lib/kubelet/plugins_registry
             type: Directory
           name: registration-dir
+        - hostPath:
+            # 'path' is where PV data is persisted on host.
+            path: /var/lib/csi-hostpath-data/
```
Comment (pohly) [inline test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml:108]: These tests are going to run in parallel, potentially on the same hosts. We can't use `/var/lib/csi-hostpath-data` on the host.

I don't think any of the mock tests rely on persisting the driver state across pod restarts. Therefore let's not use any hostpath volume. Instead pass `-statedir=/tmp/csi-hostpath-data` as parameter.

https://github.com/kubernetes-csi/csi-driver-host-path/blob/d0d44abf58958cbf2105a56da4968db2cee26291/cmd/hostpathplugin/main.go#L49

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml (hunk 1 above)
Comment (avalluri) [reply test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml]: I changed to use `--statedir` for saving the state.

File: test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml [hunk 2]
```diff
@@ -53,10 +53,17 @@ spec:
...
+          #  - https://github.com/kubernetes-csi/csi-driver-host-path/pull/269
+          #  - https://github.com/kubernetes-csi/csi-driver-host-path/pull/260          #
+          image: k8s.gcr.io/sig-storage/hostpathplugin:canary
           args:
-            - "--name=mock.storage.k8s.io"
-            - "-v=3" # enabled the gRPC call logging
+            - "--drivername=mock.storage.k8s.io"
+            - "--nodeid=$(KUBE_NODE_NAME)"
+            - "--endpoint=/csi/csi.sock"
+            - "--statedir=/csi-data-dir"
```
Comment (pohly) [inline test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml:65]: My proposal was to use `/tmp/csi-hostpath-data`, not `/csi-data-dir`. It makes the intent clearer (don't persist) and might also be a bit more efficient (tmpfs vs overlayfs with actual disk IO).

File: test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml [hunk 3]
```diff
@@ -53,10 +53,17 @@ spec:
           - mountPath: /registration
             name: registration-dir
         - name: mock
-          image: k8s.gcr.io/sig-storage/mock-driver:v4.1.0
+          # FIXME: Move to release image once the hostpath-plugin release
+          # having below chagnes;
+          #  - https://github.com/kubernetes-csi/csi-driver-host-path/pull/269
+          #  - https://github.com/kubernetes-csi/csi-driver-host-path/pull/260          #
+          image: k8s.gcr.io/sig-storage/hostpathplugin:canary
```
Comment (pohly) [inline test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml:60]: The "canary" images don't get promoted to `k8s.gcr.io/sig-storage`. You have to use `docker pull gcr.io/k8s-staging-sig-storage/hostpathplugin:canary`.

Comment (pohly) [reply test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml]: This patch enables pulling that image:
```
diff --git a/test/utils/image/manifest.go b/test/utils/image/manifest.go
index a8f60a8de85..068ef120031 100644
--- a/test/utils/image/manifest.go
+++ b/test/utils/image/manifest.go
@@ -30,18 +30,19 @@ import (
 
 // RegistryList holds public and private image registries
 type RegistryList struct {
-	GcAuthenticatedRegistry string `yaml:"gcAuthenticatedRegistry"`
-	E2eRegistry             string `yaml:"e2eRegistry"`
-	PromoterE2eRegistry     string `yaml:"promoterE2eRegistry"`
-	BuildImageRegistry      string `yaml:"buildImageRegistry"`
-	InvalidRegistry         string `yaml:"invalidRegistry"`
-	GcEtcdRegistry          string `yaml:"gcEtcdRegistry"`
-	GcRegistry              string `yaml:"gcRegistry"`
-	SigStorageRegistry      string `yaml:"sigStorageRegistry"`
-	GcrReleaseRegistry      string `yaml:"gcrReleaseRegistry"`
-	PrivateRegistry         string `yaml:"privateRegistry"`
-	SampleRegistry          string `yaml:"sampleRegistry"`
-	MicrosoftRegistry       string `yaml:"microsoftRegistry"`
+	GcAuthenticatedRegistry  string `yaml:"gcAuthenticatedRegistry"`
+	E2eRegistry              string `yaml:"e2eRegistry"`
+	PromoterE2eRegistry      string `yaml:"promoterE2eRegistry"`
+	BuildImageRegistry       string `yaml:"buildImageRegistry"`
+	InvalidRegistry          string `yaml:"invalidRegistry"`
+	GcEtcdRegistry           string `yaml:"gcEtcdRegistry"`
+	GcRegistry               string `yaml:"gcRegistry"`
+	SigStorageRegistry       string `yaml:"sigStorageRegistry"`
+	SigStorageCanaryRegistry string `yaml:"sigStorageCanaryRegistry"`
+	GcrReleaseRegistry       string `yaml:"gcrReleaseRegistry"`
+	PrivateRegistry          string `yaml:"privateRegistry"`
+	SampleRegistry           string `yaml:"sampleRegistry"`
+	MicrosoftRegistry        string `yaml:"microsoftRegistry"`
 }
 
 // Config holds an images registry, name, and version
@@ -68,18 +69,19 @@ func (i *Config) SetVersion(version string) {
 
 func initReg() RegistryList {
 	registry := RegistryList{
-		GcAuthenticatedRegistry: "gcr.io/authenticated-image-pulling",
-		E2eRegistry:             "gcr.io/kubernetes-e2e-test-images",
-		PromoterE2eRegistry:     "k8s.gcr.io/e2e-test-images",
-		BuildImageRegistry:      "k8s.gcr.io/build-image",
-		InvalidRegistry:         "invalid.com/invalid",
-		GcEtcdRegistry:          "k8s.gcr.io",
-		GcRegistry:              "k8s.gcr.io",
-		SigStorageRegistry:      "k8s.gcr.io/sig-storage",
-		PrivateRegistry:         "gcr.io/k8s-authenticated-test",
-		SampleRegistry:          "gcr.io/google-samples",
-		GcrReleaseRegistry:      "gcr.io/gke-release",
-		MicrosoftRegistry:       "mcr.microsoft.com",
+		GcAuthenticatedRegistry:  "gcr.io/authenticated-image-pulling",
+		E2eRegistry:              "gcr.io/kubernetes-e2e-test-images",
+		PromoterE2eRegistry:      "k8s.gcr.io/e2e-test-images",
+		BuildImageRegistry:       "k8s.gcr.io/build-image",
+		InvalidRegistry:          "invalid.com/invalid",
+		GcEtcdRegistry:           "k8s.gcr.io",
+		GcRegistry:               "k8s.gcr.io",
+		SigStorageRegistry:       "k8s.gcr.io/sig-storage",
+		SigStorageCanaryRegistry: "gcr.io/k8s-staging-sig-storage",
+		PrivateRegistry:          "gcr.io/k8s-authenticated-test",
+		SampleRegistry:           "gcr.io/google-samples",
+		GcrReleaseRegistry:       "gcr.io/gke-release",
+		MicrosoftRegistry:        "mcr.microsoft.com",
 	}
 	repoList := os.Getenv("KUBE_TEST_REPO_LIST")
 	if repoList == "" {
@@ -105,18 +107,19 @@ var (
 	PrivateRegistry = registry.PrivateRegistry
 
 	// Preconfigured image configs
-	dockerLibraryRegistry   = "docker.io/library"
-	e2eRegistry             = registry.E2eRegistry
-	promoterE2eRegistry     = registry.PromoterE2eRegistry
-	buildImageRegistry      = registry.BuildImageRegistry
-	gcAuthenticatedRegistry = registry.GcAuthenticatedRegistry
-	gcEtcdRegistry          = registry.GcEtcdRegistry
-	gcRegistry              = registry.GcRegistry
-	sigStorageRegistry      = registry.SigStorageRegistry
-	gcrReleaseRegistry      = registry.GcrReleaseRegistry
-	invalidRegistry         = registry.InvalidRegistry
-	sampleRegistry          = registry.SampleRegistry
-	microsoftRegistry       = registry.MicrosoftRegistry
+	dockerLibraryRegistry    = "docker.io/library"
+	e2eRegistry              = registry.E2eRegistry
+	promoterE2eRegistry      = registry.PromoterE2eRegistry
+	buildImageRegistry       = registry.BuildImageRegistry
+	gcAuthenticatedRegistry  = registry.GcAuthenticatedRegistry
+	gcEtcdRegistry           = registry.GcEtcdRegistry
+	gcRegistry               = registry.GcRegistry
+	sigStorageRegistry       = registry.SigStorageRegistry
+	sigStorageCanaryRegistry = registry.SigStorageCanaryRegistry
+	gcrReleaseRegistry       = registry.GcrReleaseRegistry
+	invalidRegistry          = registry.InvalidRegistry
+	sampleRegistry           = registry.SampleRegistry
+	microsoftRegistry        = registry.MicrosoftRegistry
 
 	imageConfigs, originalImageConfigs = initImageConfigs()
 )
@@ -393,6 +396,8 @@ func ReplaceRegistryInImageURL(imageURL string) (string, error) {
 		registryAndUser = gcRegistry
 	case "k8s.gcr.io/sig-storage":
 		registryAndUser = sigStorageRegistry
+	case "gcr.io/k8s-staging-sig-storage":
+		registryAndUser = sigStorageCanaryRegistry
 	case "gcr.io/k8s-authenticated-test":
 		registryAndUser = PrivateRegistry
 	case "gcr.io/google-samples":
```

I wonder whether we should add that permanently and not just while experimenting with a canary image. I remember having to add and remove this already a few times.

@msau42 any thoughts on this?

@avalluri : put it into a separate commit for now, please.

Comment (avalluri) [reply test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml]: Done.

File: test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml (hunk 2 above)
Comment (avalluri) [reply test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml]: Changed statedir to /tmp/csi-hostpath-data, as suggested.

**Pull request closed without merging**